### ISSUE TITLE
Add incomplete gamma functions

### ISSUE THREAD

**Issue by hawkinsp** [google/jax#479]
i.e., `scipy.special.gammainc` and `scipy.special.gammaincc`.

**Comment by CamDavidsonPilon** [comment]
@hawkinsp what are possible steps needed to implement this? I'm happy to try my hand at it.

**Comment by mattjj** [comment]
Awesome, thanks for your interest!

AIUI there are basically two ways to add functions like these:
1. implement them ourselves in terms of lax primitives, like how we implement, say, [`expit`](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jax/scipy/special.py#L69-L72) or [`logaddexp`](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jax/numpy/lax_numpy.py#L440-L444)
2. find an implementation in the [C++ XLA client math library](https://github.com/tensorflow/tensorflow/blob/ed14b5da3f7e9105bea3f7314706d04a68969edc/tensorflow/compiler/xla/client/lib/math.cc) and add it to JAX as a primitive in lax.py, like we do with [`lax.digamma`](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jax/lax/lax.py#L173-L175) [here](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jax/lax/lax.py#L1618).

Interestingly, these two approaches are pretty similar: the XLA client libraries just implement functions in terms of XLA:HLO, and that's exactly what we get when we write functions at the Python level in terms of the primitives in lax.py. That's because the lax.py primitives are essentially 1:1 with XLA HLO (or in a few cases client library functions), so they translate to the same XLA computations. In general I think it's easier and nicer to implement things in Python in terms of lax.py, but on the other hand for trickier numerical computations it can be good to reuse the effort in the XLA client library, and then we also get to share any updates or fixes across XLA frontends.

My guess is that (1) is the best approach to start with. On that front, I just noticed that @fehiepsi added some [calculations involving incomplete gamma expressions in random.py](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jax/random.py#L595-L620), so that might be a good source of clues.

I think the standard starting place is to check whether the XLA client library implements these things for us (in terms of other XLA HLO computations) or if we need to implement them ourselves. [Here's part of the XLA client math library](https://github.com/tensorflow/tensorflow/blob/ed14b5da3f7e9105bea3f7314706d04a68969edc/tensorflow/compiler/xla/client/lib/math.cc#L353) that seems relevant, though I'm not sure if it's got incomplete gamma stuff yet. I'd recommend checking there and in any nearby source files.

**Comment by mattjj** [comment]
Whoops, hit the wrong button :)

**Comment by CamDavidsonPilon** [comment]
From what I understand, the way to implement this is to natively add algorithms for computing gammainc. After looking at how scipy computes gammainc, it's _not_ a trivial implementation (ex: involves hours-to-days of compute time to determine precomputed coefficients). That makes me nervous. I feel we should be leaning strongly on well-known implementations¹. However, maybe the scope of Jax is larger than I am imagining, and Jax/XLA does need a native implementation. 


¹ If Jax/XLA does have a native implementation, does this mean that the Python/Scipy/Numpy code is really just a DSL? Example: in the future, when I call `jax.scipy.special.gammainc`, my code path is being diverted _from_ the scipy implementation _to_ a Jax/XLA implementation. What happens when `jax.scipy.special.gammainc` gives different results from `scipy.special.gammainc`?

**Comment by mattjj** [comment]
> That makes me nervous. I feel we should be leaning strongly on well-known implementations.

We could set up JAX to call into existing platform-specific implementations, like it does with linear algebra routines on CPU and GPU. Setting up those low-level bindings requires a bit more low-level work though.

> If Jax/XLA does have a native implementation, does this mean that the Python/Scipy/Numpy code is really just a DSL?

It's useful in general to think of how we write array-oriented programs in terms of an embedded DSL. That is, NumPy itself embeds a language in Python. We clearly feel the difference between staying in the embedded NumPy language and popping up to the meta-programming Python layer whenever we do something like write Python loops and thus fall off a performance cliff.

> my code path is being diverted from the scipy implementation to a Jax/XLA implementation

Indeed, that's a pretty foundational design of JAX: we're not using the same implementations as NumPy/SciPy because, for example, we want to run on GPUs and TPUs.

> What happens when jax.scipy.special.gammainc gives different results from scipy.special.gammainc?

JAX isn't a SciPy (or NumPy) replacement, and it doesn't guarantee identical numerical results. We do test extensively against NumPy and SciPy, it's only up to some numerical tolerances.

As I think your comment is getting at, we're using the NumPy (and SciPy) API as a way to express numerical programs, like other libraries do (like [dask](https://dask.org) for example). Here's a long snippet from [NEP18](https://www.numpy.org/neps/nep-0018-array-function-protocol.html):

> We propose the `__array_function__` protocol, to allow arguments of NumPy functions to define how that function operates on them. This will allow using NumPy as a high level API for efficient multi-dimensional array operations, even with array implementations that differ greatly from `numpy.ndarray`.
>
> NumPy’s high level ndarray API has been implemented several times outside of NumPy itself for different architectures, such as for GPU arrays (CuPy), Sparse arrays (scipy.sparse, pydata/sparse) and parallel arrays (Dask array) as well as various NumPy-like implementations in the deep learning frameworks, like TensorFlow and PyTorch.
>
> Similarly there are many projects that build on top of the NumPy API for labeled and indexed arrays (XArray), automatic differentiation (Autograd, Tangent), masked arrays (numpy.ma), physical units (astropy.units, pint, unyt), etc. that add additional functionality on top of the NumPy API. Most of these project also implement a close variation of NumPy’s level high API.

What do you think?

**Comment by mattjj** [comment]
To give a bit more detail on how to set up platform-specific implementations, in terms of low-level code [here's how we call into LAPACK routines on CPU](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jaxlib/lapack.pyx) by [grabbing the function pointers SciPy provides](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jaxlib/lapack.pyx#L28-L33) and [registering XLA:CPU CustomCalls into them](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jaxlib/lapack.pyx#L382). Then we just [set up a platform-specific translation rule for linalg functions](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jax/lax_linalg.py#L136-L156), which ultimately is what we [call into from our NumPy layer](https://github.com/google/jax/blob/61e52ab2b5f3a7f593e15301f54ad142b8d44718/jax/numpy/linalg.py#L53).

The [SciPy Cython API for special functions](https://docs.scipy.org/doc/scipy/reference/special.cython_special.html) also includes `double gammainc(double, double)` and all the other stuff we need, so we can follow the same pattern for CPU! For GPU we'd need to find the right CUDA routines.

**Comment by CamDavidsonPilon** [comment]
Thanks for the detailed response @mattjj. That makes a lot of sense, specifically the numpy analogy.

Given what I now know about implementing a gammainc function, I believe it's beyond my capabilities/expertise, so I'll leave it for someone else. Hopefully this issue's discussion gives them a better starting point!

**Comment by hawkinsp** [comment]
This was fixed by @srvasude in PR #2064 .
